File: HappyPuppy/C#/Repository.cs
using System;
using System.Collections.Generic;

public interface HappyPuppyRepository
{
    IEnumerable<string> RetrieveOwners();
    IEnumerable<string> RetrievePuppies();
    void RemovePuppy(string name);
    void AddOwner(string name);
}

public class ListBasedRepository: HappyPuppyRepository
{
        private Dictionary<string, List<string>> database;

        public ListBasedRepository(Dictionary<string, List<string>> database)
        {
            this.database = database;
        }

        public IEnumerable<string> RetrieveOwners()
        {
            return database["owners"];
        }

        public IEnumerable<string> RetrievePuppies()
        {
            return database["puppies"];
        }

        public void RemovePuppy(string name)
        {
            database["puppies"].Remove(name);
        }

        public void AddOwner(string name)
        {
            database["owners"].Add(name);
        }
}

File: HappyPuppy/Java/Repository.java
import java.util.*;

public class Repository
{
    public interface HappyPuppyRepository
    {
        Iterable<String> RetrieveOwners();
        Iterable<String> RetrievePuppies();
        void RemovePuppy(String name);
        void AddOwner(String name);
    }

    public static class ListBasedRepository implements HappyPuppyRepository
    {
            private Map<String, List<String>> database;
    
            public ListBasedRepository(Map<String, List<String>> database)
            {
                this.database = database;
            }
    
            public Iterable<String> RetrieveOwners()
            {
                return database.get("owners");
            }
    
            public Iterable<String> RetrievePuppies()
            {
                return database.get("puppies");
            }
    
            public void RemovePuppy(String name)
            {
                database.get("puppies").remove(name);
            }
    
            public void AddOwner(String name)
            {
                database.get("owners").add(name);
            }
    }
}
